### PR TITLE
feat(doc): Add design doc for batch local execution mode

### PR DESCRIPTION
**Pull request by liurenjie1024**
## What's changed and what's your intention?

RT

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

### CODE REVIEW

File: docs/batch-local-execution-mode.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+## Example 1: select a from t where b in (1, 2, 3, 4)
+
+Let’s use above sql as example:
+
+
+![Example 1](./images/batch-local-execution-mode/example1.svg)
+
+The key changes from distributed mode:
+
+1. The exchange executor will be executed by directly local query execution, not by distributed scheduler. This means that we no longer have async execution/monitoring, etc.
```
Comment (lmatz) [inline docs/batch-local-execution-mode.md:32]: nit: "by directy" -> "directly by"

File: docs/batch-local-execution-mode.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+![Example 2](./images/batch-local-execution-mode/example2.svg)
+
+As explained above, the lookup join/exchange phase will be executed directly on frontend. The pushdown(filter/table, both the build and probe side) will be issued by executors rather than scheduler.
+
+### Optimization/Scheduling
+
+The overall process will be quite similar to distributed processing, but with a little difference:
+
+1. We only use heuristic optimizater for it, and only a limit set of rules will be applied.
```
Comment (lmatz) [inline docs/batch-local-execution-mode.md:48]: nit: "a limit set" -> "a limited set"

File: docs/batch-local-execution-mode.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+As mentioned in first paragraph, the main use case for local execution mode is determined(dashboard/reporting), so currently we just expose a session configuration(`query_mode`) to user. In future we may use optimizer to determined it, but it depends on requirement.
+
+### RPC execution in local mode
+
+In distributed mode we have several steps to execute compute task and fetch results:
+
+There are some problems with above process in local mode:
+
+1. We need at least two rpcs to fetch task execution result, this increases query overhead
+2. We have task lifecycle management apis, this is unnecessary for local mode.
```
Comment (lmatz) [inline docs/batch-local-execution-mode.md:67]: Does "apis" mean "APIs"?

**Review by lmatz — APPROVED** [review]

File: docs/batch-local-execution-mode.md (hunk 3 above)
Comment (liurenjie1024) [reply docs/batch-local-execution-mode.md:67]: Yes.

File: docs/batch-local-execution-mode.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
+# Design: Local execution mode for online serving queries
+
+## Background
+
+The query engine of risingwave is expected to support two kinds of query: highly concurrent point query and adhoc query.
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:5]: -> The query engine of RisingWave supports two types of queries: highly concurrent point queries and ad-hoc queries.

File: docs/batch-local-execution-mode.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+Our distributed query processing engine is designed for complex adhoc query, and it can’t meet the latency/qps 
+requirement of point query, and in this article we introduce local execution mode for point query.
+
+## Design
+
+![Frontend Flow](./images/batch-local-execution-mode/frontend-flow.svg)
+
+## Example 1: select a from t where b in (1, 2, 3, 4)
+
+Let’s use above sql as example:
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:24]: -> the above SQL as an example:

File: docs/batch-local-execution-mode.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
+# Design: Local execution mode for online serving queries
+
+## Background
+
+The query engine of risingwave is expected to support two kinds of query: highly concurrent point query and adhoc query.
+The characteristics of these two different kinds of queries are summarized as followings:
+
+|	|Point Query	|Adhoc query	|
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:7]: Use the plural form "queries" when possible.

File: docs/batch-local-execution-mode.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+|---	|---	|---	|
+|Latency	|several ms	|ms to minutes	|
+|QPS	|1000 ~ 100000	|100 ~ 1000	|
+|SQL	|Simple	|Arbitrary complex	|
+|Result Set	|Small	|Small, Medium, Large	|
+|Use Case	|Dashboard	|Engineers, Data Scientist	|
+
+Our distributed query processing engine is designed for complex adhoc query, and it can’t meet the latency/qps 
+requirement of point query, and in this article we introduce local execution mode for point query.
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:16]: -> ... point queries. In this article, we'll introduce the local execution mode for point queries.

File: docs/batch-local-execution-mode.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+The query engine of risingwave is expected to support two kinds of query: highly concurrent point query and adhoc query.
+The characteristics of these two different kinds of queries are summarized as followings:
+
+|	|Point Query	|Adhoc query	|
+|---	|---	|---	|
+|Latency	|several ms	|ms to minutes	|
+|QPS	|1000 ~ 100000	|100 ~ 1000	|
+|SQL	|Simple	|Arbitrary complex	|
+|Result Set	|Small	|Small, Medium, Large	|
+|Use Case	|Dashboard	|Engineers, Data Scientist	|
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:13]: I feel listing Engineers, Data Scientist side-by-side with Dashboard isn't an accurate categorization. Maybe say "Use scenarios or target audiences".

File: docs/batch-local-execution-mode.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+![Frontend Flow](./images/batch-local-execution-mode/frontend-flow.svg)
+
+## Example 1: select a from t where b in (1, 2, 3, 4)
+
+Let’s use above sql as example:
+
+
+![Example 1](./images/batch-local-execution-mode/example1.svg)
+
+The key changes from distributed mode:
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:29]: -> from the distributed mode:

File: docs/batch-local-execution-mode.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+1. The exchange executor will be executed directly by local query execution, not by distributed scheduler. This means that we no longer have async execution/monitoring, etc.
+2. The rpc is issued by exchange executor directly, not by scheduler.
+
+## Example2: SELECT pk, t1.a, t1.fk, t2.b FROM t1, t2 WHERE t1.fk = t2.pk AND t1.pk = 114514
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:34]: Add a space between "Example" and "2".

File: docs/batch-local-execution-mode.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+![Example 2](./images/batch-local-execution-mode/example2.svg)
+
+As explained above, the lookup join/exchange phase will be executed directly on frontend. The pushdown(filter/table, both the build and probe side) will be issued by executors rather than scheduler.
+
+### Optimization/Scheduling
+
+The overall process will be quite similar to distributed processing, but with a little difference:
+
+1. We only use heuristic optimizater for it, and only a limited set of rules will be applied.
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:47]: optimizater -> optimizer

File: docs/batch-local-execution-mode.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+2. No scheduler will involve, and the physical plan is executed in current thread(coroutine) immeidately.
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:48]: As I understand it, there is only one scheduler. Correct? If so, I would reword the sentence to be:
The scheduler will not be involved, and the physical plan is executed in the current thread (coroutine) immediately.

File: docs/batch-local-execution-mode.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+### Monitoring/Management
+
+Local execution mode will not go through query management mentioned in  batch query manager to reduce latency as 
+much as possible.
+
+### How to switch between local/distributed execution mode
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:55]: -> ... local and distributed execution modes?
Please add a question mark at the end.

File: docs/batch-local-execution-mode.md [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+As mentioned in first paragraph, the main use case for local execution mode is determined(dashboard/reporting), so currently we just expose a session configuration(`query_mode`) to user. In future we may use optimizer to determined it, but it depends on requirement.
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:57]: -> in the first paragraph
-> ... use optimizer to determine it if required.

File: docs/batch-local-execution-mode.md [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+
+### RPC execution in local mode
+
+In distributed mode we have several steps to execute compute task and fetch results:
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:61]: -> In the distributed mode...execute a compute task...

File: docs/batch-local-execution-mode.md [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+There are some problems with above process in local mode:
+
+1. We need at least two rpcs to fetch task execution result, this increases query overhead
+2. We have task lifecycle management APIs, this is unnecessary for local mode.
+3. We may need to add several new apis for task monitoring/failure detection
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:67]: Write as API or APIs consistently.

File: docs/batch-local-execution-mode.md [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+
+
+For local mode we will add a new rpc api:
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:70]: Refer the local mode as "the local mode".

File: docs/batch-local-execution-mode.md [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+
+message ExecuteResponse {
+  common.Status status = 1;
+  data.DataChunk record_batch = 2;
+}
+```
+
+This is quite similar to distributed execution api, but with some differences:
+
+1. Response is returned directly in rpc call, without the need of another call
```
Comment (hengm3467) [inline docs/batch-local-execution-mode.md:88]: End this item with a period.

**Review by hengm3467 — COMMENTED** [review]
Added some minor stylish comments. Otherwise looks good.

**Comment by liurenjie1024** [comment]
> Added some minor stylish comments. Otherwise looks good. 

Thanks for the review.
